state_synchronizer/src/peer_manager.rs

```diff
@@ -94,28 +101,51 @@ impl PeerManager {
         if let Some(peer_info) = self.peers.get_mut(peer_id) {
             match update_type {
                 PeerScoreUpdateType::Success => {
-                    let new_score = peer_info.score + 1.0;
-                    peer_info.score = new_score.min(MAX_SCORE);
+                    if peer_info.score < MAX_SCORE {
+                        let new_score = peer_info.score + 1.0;
+                        peer_info.score = new_score.min(MAX_SCORE);
+                        self.weighted_index = self.compute_weighted_index();
+                    }
                 }
                 PeerScoreUpdateType::InvalidChunk => {
-                    let new_score = peer_info.score * 0.8;
-                    peer_info.score = new_score.max(MIN_SCORE);
+                    if peer_info.score > MIN_SCORE {
+                        let new_score = peer_info.score * 0.8;
+                        peer_info.score = new_score.max(MIN_SCORE);
+                        self.weighted_index = self.compute_weighted_index();
+                    }
                 }
             }
         }
     }
 
-    pub fn pick_peer(&mut self) -> Option<(PeerId, StateSynchronizerSender)> {
+    fn compute_weighted_index(&self) -> Option<WeightedIndex<f64>> {
         let active_peers = self.get_active_upstream_peers();
-        debug!("[state sync] (pick_peer) state: {:?}", self.peers);
-
         if !active_peers.is_empty() {
             let weights: Vec<_> = active_peers
                 .iter()
                 .map(|(_, peer_info)| peer_info.score)
                 .collect();
+            match WeightedIndex::new(&weights) {
+                Ok(weighted_index) => {
+                    return Some(weighted_index);
+                }
+                Err(e) => {
+                    error!(
+                        "[state sync] (pick_peer) invalid weighted index distribution, {:?}",
+                        e
+                    );
+                }
+            }
+        }
+        None
+    }
+
+    pub fn pick_peer(&mut self) -> Option<(PeerId, StateSynchronizerSender)> {
+        let active_peers = self.get_active_upstream_peers();
+        debug!("[state sync] (pick_peer) state: {:?}", self.peers);
 
-            if let Ok(weighted_index) = WeightedIndex::new(&weights) {
+        if !active_peers.is_empty() {
+            if let Some(weighted_index) = self.weighted_index.clone() {
```

Comment:
can we sample without cloning the weighted index?

Reply:
yeah, isn't clone diminishes whole optimization?